### DISCUSSION TITLE
can you import glyphs from a font into different glyph places into new font

### DISCUSSION THREAD

**Discussion by waykyi** [fontforge/fontforge#4791] [Q&A]
I have fonts that were written all glyphs in the English glyph places. Now our language is supported in Unicode and the glyphs are assigned into way down in the glyph places. Do you have utility or import function that can read from old fonts and place them into appropriate places in Unicode? Thus creating a new font with glyphs are placed into assigned places into Unicode. Please help point me to such program or app out there so that it can be done properly and quickly. Thanks

**Comment by ctrlcctrlv** [comment]
You could probably do that with a FontForge script, yes, but you'd need to write it.

First you need to come up with a mapping from the ASCII codepoints to the Unicode codepoints, that would look something like:

```python3
MAP = {
  ord("A"): 0x2828,
  ord("B"): 0x2829,
  # …etc…
}
```

Then you need to…

```python3
import fontforge
import sys
font = fontforge.open(sys.argv[1])
for glyph in font.glyphs():
  glyph.unicode = MAP[glyph.unicode]
font.generate(…)
```

Not difficult.

**Reply by waykyi** [reply]
Wow, this is quick and appreciate you very much. I hope I can try & error and able to do it. This is great help and if anything success or stuck, I will let you know in here. 

Just one more question is: you have to open the old font first in FF and run the Python script with .py extension, right? I have no experience in Python but I knew some Object Oriented Programming. Thanks

**Reply by ctrlcctrlv** [reply]
If you have the script saved to a file and the Python interpreter you call has access to the fontforge module, there's no need to open the font in FontForge's GUI.

Whether or not this is true by default depends on your OS. It is true most of the time on Linux, as long as you're using distribution packages.

![image](https://user-images.githubusercontent.com/838783/129497717-334c2b4d-9d12-4401-8b1a-c0977bb7b6d0.png)

On Windows you may have to specifically run the `ffpython.exe` which is distributed along with FontForge on your script.

**Reply by waykyi** [reply]
Sorry, I just noticed that your code was in FF scripting. I will try it tonight. Any pointer to FF scripting manual and samples? Seem like it is very easier than Python. Thanks

**Reply by ctrlcctrlv** [reply]
My code is not in "FF scripting", it's in Python. The FontForge internal scripting language is deprecated and not as featureful as Python. It's recommended to use Python.

**Reply by waykyi** [reply]
I was able to do some FF script working but there was an error "ScriptDlg: 12 End of the file found in If ff_statement". But glyphs are copied to new file. Now, I want to able to do a loop with all glyphs copied into new file, how do I do that? Can I do two arrays with one for old glyphs numbers and the other for new glyphs numbers where I want to paste them. And one more thing is, can I do without open/close again and again both font files? If I can switch between old and new file, what is the function on that? 
Here is my script:
myFile = ( "c://My Documents//Fontforge//Font Designs//Test//MyTest.sfd" )
if (FileAccess(myFile)==0)
Open(myFile);
Select(33);
if (WorthOutputting())
Select(33,34,35,36,37,38,39,40,41,42);
Copy();
Close();
New();
Select(0x1006,0x1007,0x1008,0x1009,0x100A,0x100B,0x100C,0x100D,0x100E,0x100F);
Paste();
endif

Your direction is very much appreciated. Thanks

![error](https://user-images.githubusercontent.com/73824794/129726804-207cab60-3b0d-4f05-96a7-7b55bc191ec3.png)

**Reply by waykyi** [reply]
I think I got them in FF script. Just still need switching open files in FF, which function or how to switch one file to another. Can you help? Thanks
